File: src/src_core/fmg/core/img/MosaicsSkillImg.java
package fmg.core.img;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import fmg.common.geom.PointDouble;
import fmg.common.geom.util.FigureHelper;
import fmg.data.controller.types.ESkillLevel;

/**
 * Representable {@link fmg.data.controller.types.ESkillLevel} as image
 *
 * @param <TImage> plaform specific image
 */
public abstract class MosaicsSkillImg<TImage> extends RotatedImg<ESkillLevel, TImage> {

   protected MosaicsSkillImg(ESkillLevel skill) { super(skill); }

   public ESkillLevel getMosaicSkill() { return getEntity(); }
   public void setMosaicSkill(ESkillLevel value) { setEntity(value); }

   protected Stream<Stream<PointDouble>> getCoords() {
      double sq = Math.min( // size inner square
            getWidth()  - getPadding().getLeftAndRight(),
            getHeight() - getPadding().getTopAndBottom());
      double r1 = sq/7; // external radius
      double r2 = sq/12; // internal radius
      int ordinal = getMosaicSkill().ordinal();
      int rays = 5 + ordinal; // rays count
      int stars = 4 + ordinal; // number of stars on the perimeter of the circle
      double[] angle = { getRotateAngle() };
      double starAngle = 360.0/stars;
      final PointDouble center = new PointDouble(getWidth() / 2.0, getHeight() / 2.0);
      final PointDouble zero = new PointDouble(0, 0);
      Stream<Stream<PointDouble>> res = IntStream.range(0, stars)
            .mapToObj(st -> {
               // (un)comment next line to view result changes...
               angle[0] = Math.sin(FigureHelper.toRadian(angle[0]/4))*angle[0]; // accelerate / ускоряшка..

               // adding offset
               PointDouble offset = FigureHelper.getPointOnCircle(sq / 3, angle[0] + (st * starAngle), zero);
               PointDouble centerStar = new PointDouble(center.x + offset.x, center.y + offset.y);

               Stream<PointDouble> points = (getMosaicSkill() == ESkillLevel.eCustom)
                     ? FigureHelper.getRegularPolygonCoords(3 + (st % 4), r1, centerStar, -angle[0])
                     : FigureHelper.getRegularStarCoords(rays, r1, r2, centerStar, -angle[0]);
               return points;
            });
      List<Stream<PointDouble>> resL = res.collect(Collectors.toList());
      Collections.reverse(resL); // reverse stars, to draw the first star of the latter. (pseudo Z-order). (un)comment line to view result changes...
      return resL.stream();
   }

}

File: src/src_core/fmg/core/img/MosaicsSkillImg.cs
using System;
using System.Linq;
using System.Collections.Generic;
using fmg.common.geom;
using fmg.common.geom.util;
using fmg.data.controller.types;

namespace fmg.core.img {

   /// <summary> Representable <see cref="ESkillLevel"/> as image </summary>
   /// <typeparam name="TImage">plaform specific image</typeparam>
   public abstract class MosaicsSkillImg<TImage> : RotatedImg<ESkillLevel, TImage>
      where TImage : class
   {
      protected MosaicsSkillImg(ESkillLevel group)
         : base(group) {}

      public ESkillLevel MosaicSkill => Entity;

      protected IEnumerable<IEnumerable<PointDouble>> GetCoords() {
         double sq = Math.Min( // size inner square
            Width - Padding.LeftAndRight,
            Height - Padding.TopAndBottom);
         var r1 = sq/7; // external radius
         var r2 = sq/12; // internal radius
         var ordinal = MosaicSkill.Ordinal();
         var rays = 5 + ordinal; // rays count
         var stars = 4 + ordinal; // number of stars on the perimeter of the circle
         var angle = RotateAngle;
         var starAngle = 360.0/stars;
         var center = new PointDouble(Width / 2.0, Height / 2.0);
         var zero = new PointDouble(0, 0);
         return Enumerable.Range(0, stars).Select(st => {
            // (un)comment next line to view result changes...
            angle = Math.Sin((angle/4).ToRadian())*angle; // accelerate / ускоряшка..

            // adding offset
            var offset = FigureHelper.GetPointOnCircle(sq/3, angle + st*starAngle, zero);
            var centerStar = new PointDouble(center.X + offset.X, center.Y + offset.Y);

            return (MosaicSkill == ESkillLevel.eCustom)
               ? FigureHelper.GetRegularPolygonCoords(3 + st % 4, r1, centerStar, -angle)
               : FigureHelper.GetRegularStarCoords(rays, r1, r2, centerStar, -angle);

         })
         .Reverse(); // reverse stars, to draw the first star of the latter. (pseudo Z-order). (un)comment line to view result changes...
      }

   }

}
